File: packages/python/src/jev_moderation/evaluation.py
"""Labeled evaluations use the same moderator as production, with bounded concurrency."""

from __future__ import annotations

import asyncio
import math
import time
from typing import TypedDict

from typing_extensions import NotRequired

from .core import (
    Action,
    AsyncModerator,
    ConfigurationError,
    ModerationError,
    ModerationResult,
    PolicyConfig,
    QualityGates,
    TestCase,
    Usage,
    decide_result,
    digest,
    inference_hash,
    mock_response,
    resolve_policy,
    validate_cases,
    validate_schema,
)


class EvaluationError(TypedDict):
    code: str
    message: str


class CaseResult(TypedDict):
    id: str
    expectedAction: Action
    latencyMs: float
    result: NotRequired[ModerationResult]
    error: NotRequired[EvaluationError]


class FilterMetrics(TypedDict):
    tp: int
    fp: int
    tn: int
    fn: int
    precision: float | None
    recall: float | None
    falsePositiveIds: list[str]
    falseNegativeIds: list[str]


class LatencyMetrics(TypedDict):
    mean: float
    p95: float


class EvaluationReport(TypedDict):
    schemaVersion: int
    policyHash: str
    policyVersion: str
    inferenceHash: str
    datasetHash: str
    requestedModel: str
    total: int
    completed: int
    errors: int
    accuracy: float
    reviewRate: float
    confusion: dict[Action, dict[Action, int]]
    filters: dict[str, FilterMetrics]
    latencyMs: LatencyMetrics
    usage: Usage
    failedCaseIds: list[str]
    cases: list[CaseResult]
    gateFailures: list[str]
    passed: bool
    reused: bool


class ComparisonReport(TypedDict):
    left: EvaluationReport
    right: EvaluationReport
    reused: bool
    changedCaseIds: list[str]
    passed: bool


def validate_gates(gates: QualityGates, policy: dict) -> None:
    validate_schema("gates", gates)
    ids = {f["id"] for f in policy["filters"]}
    for id in gates.get("filters", {}):
        if id not in ids:
            raise ConfigurationError(f"Unknown gate filter: {id}")


def summarize(
    policy: dict,
    model: str,
    cases: list[TestCase],
    rows: list[dict],
    gates: QualityGates | None = None,
    reused: bool = False,
) -> EvaluationReport:
    gates = {"minAccuracy": 1} if gates is None else gates
    validate_gates(gates, policy)
    actions = ["allow", "review", "block"]
    confusion = {a: {b: 0 for b in actions} for a in actions}
    filters = {
        f["id"]: dict(
            tp=0,
            fp=0,
            tn=0,
            fn=0,
            precision=None,
            recall=None,
            falsePositiveIds=[],
            falseNegativeIds=[],
        )
        for f in policy["filters"]
    }
    usage, failed = dict(input_tokens=0, output_tokens=0), []
    correct = reviews = errors = 0
    for c, row in zip(cases, rows):
        if "result" not in row:
            errors += 1
            failed.append(row["id"])
            continue
        r = row["result"]
        confusion[c["expectedAction"]][r["action"]] += 1
        if r["action"] == c["expectedAction"]:
            correct += 1
        else:
            failed.append(row["id"])
        if r["action"] == "review":
            reviews += 1
        for key in usage:
            usage[key] += r["usage"][key]
        for id, truth in c.get("labels", {}).items():
            predicted, m = r["filters"][id]["action"] != "allow", filters[id]
            if predicted and truth:
                m["tp"] += 1
            elif predicted:
                m["fp"] += 1
                m["falsePositiveIds"].append(row["id"])
            elif truth:
                m["fn"] += 1
                m["falseNegativeIds"].append(row["id"])
            else:
                m["tn"] += 1
            if predicted != truth and row["id"] not in failed:
                failed.append(row["id"])
    for m in filters.values():
        m["precision"] = m["tp"] / (m["tp"] + m["fp"]) if m["tp"] + m["fp"] else None
        m["recall"] = m["tp"] / (m["tp"] + m["fn"]) if m["tp"] + m["fn"] else None
    durations = sorted(row["latencyMs"] for row in rows)
    accuracy, review_rate, gate_failures = correct / len(cases), reviews / len(cases), []
    if "minAccuracy" in gates and accuracy < gates["minAccuracy"]:
        gate_failures.append("minAccuracy")
    if "maxReviewRate" in gates and review_rate > gates["maxReviewRate"]:
        gate_failures.append("maxReviewRate")
    for id, gate in gates.get("filters", {}).items():
        for key, metric in [("minPrecision", "precision"), ("minRecall", "recall")]:
            if key in gate and (filters[id][metric] is None or filters[id][metric] < gate[key]):
                gate_failures.append(f"{id}.{key}")
    return dict(
        schemaVersion=1,
        policyHash=digest(policy),
        policyVersion=policy["version"],
        inferenceHash=inference_hash(policy, model),
        datasetHash=digest(cases),
        requestedModel=model,
        total=len(cases),
        completed=len(cases) - errors,
        errors=errors,
        accuracy=accuracy,
        reviewRate=review_rate,
        confusion=confusion,
        filters=filters,
        latencyMs={
            "mean": sum(durations) / len(rows),
            "p95": durations[math.ceil(len(durations) * 0.95) - 1],
        },
        usage=usage,
        failedCaseIds=failed,
        cases=rows,
        gateFailures=gate_failures,
        passed=errors == 0 and not gate_failures,
        reused=reused,
    )


async def evaluate_async(
    policy: PolicyConfig,
    cases: list[TestCase],
    *,
    moderator: AsyncModerator | None = None,
    concurrency: int = 4,
    gates: QualityGates | None = None,
) -> EvaluationReport:
    resolved = resolve_policy(policy)
    validate_cases(cases, resolved)
    validate_gates({"minAccuracy": 1} if gates is None else gates, resolved)
    if (
        isinstance(concurrency, bool)
        or not isinstance(concurrency, int)
        or not 1 <= concurrency <= 64
    ):
        raise ConfigurationError("concurrency must be an integer from 1 to 64")
    owned = moderator is None
    moderator = moderator or AsyncModerator(policy=policy)
    if digest(resolved) != digest(moderator.policy):
        raise ConfigurationError("Moderator and evaluation policies must match")
    rows = [None] * len(cases)
    next_index = 0

    async def worker():
        nonlocal next_index
        while next_index < len(cases):
            index = next_index
            next_index += 1
            c = cases[index]
            start = time.perf_counter()
            row = dict(id=c["id"], expectedAction=c["expectedAction"], latencyMs=0)
            try:
                row["result"] = await moderator.moderate(c["message"], c.get("history"))
            except Exception as e:
                row["error"] = dict(
                    code=e.code if isinstance(e, ModerationError) else "execution",
                    message=str(e) if isinstance(e, ModerationError) else "Evaluation failed",
                )
            row["latencyMs"] = (time.perf_counter() - start) * 1000
            rows[index] = row

    tasks = [asyncio.create_task(worker()) for _ in range(min(concurrency, len(cases)))]
    try:
        await asyncio.gather(*tasks)
        return summarize(resolved, moderator.model, cases, rows, gates)
    finally:
        for task in tasks:
            if not task.done():
                task.cancel()
        await asyncio.gather(*tasks, return_exceptions=True)
        if owned:
            await moderator.close()


def evaluate(policy: PolicyConfig, cases: list[TestCase], **options) -> EvaluationReport:
    """Synchronous evaluation entrypoint. In async applications use evaluate_async."""
    return asyncio.run(evaluate_async(policy, cases, **options))


def rethreshold(
    report: dict, policy: PolicyConfig, cases: list[TestCase], gates: QualityGates | None = None
) -> EvaluationReport:
    resolved = resolve_policy(policy)
    validate_cases(cases, resolved)
    if (
        report["inferenceHash"] != inference_hash(resolved, report["requestedModel"])
        or report["datasetHash"] != digest(cases)
        or len(report["cases"]) != len(cases)
    ):
        raise ConfigurationError(
            "Replay requires identical questions, requested model, and dataset"
        )
    rows = []
    for c, row in zip(cases, report["cases"]):
        if row["id"] != c["id"]:
            raise ConfigurationError("Replay case order mismatch")
        if "result" not in row:
            rows.append({**row, "latencyMs": 0})
            continue
        r = row["result"]
        raw = mock_response({id: f["probability"] for id, f in r["filters"].items()}, r["model"])
        rows.append(
            dict(
                id=row["id"],
                expectedAction=c["expectedAction"],
                latencyMs=0,
                result=decide_result(resolved, raw),
            )
        )
    return summarize(resolved, report["requestedModel"], cases, rows, gates, True)


async def compare_async(
    left: PolicyConfig,
    right: PolicyConfig,
    cases: list[TestCase],
    *,
    left_moderator: AsyncModerator | None = None,
    right_moderator: AsyncModerator | None = None,
    concurrency: int = 4,
    gates: QualityGates | None = None,
) -> ComparisonReport:
    pa, pb = resolve_policy(left), resolve_policy(right)
    for p in (pa, pb):
        validate_cases(cases, p)
        validate_gates({"minAccuracy": 1} if gates is None else gates, p)
    a, b = (
        left_moderator or AsyncModerator(policy=left),
        right_moderator or AsyncModerator(policy=right),
    )
    if digest(a.policy) != digest(pa) or digest(b.policy) != digest(pb):
        raise ConfigurationError("Moderator and comparison policies must match")
    try:
        first = await evaluate_async(left, cases, moderator=a, concurrency=concurrency, gates=gates)
        second = (
            rethreshold(first, right, cases, gates)
            if inference_hash(pa, a.model) == inference_hash(pb, b.model)
            else await evaluate_async(
                right, cases, moderator=b, concurrency=concurrency, gates=gates
            )
        )
        changed = [
            c["id"]
            for c, x, y in zip(cases, first["cases"], second["cases"])
            if x.get("result", {}).get("action") != y.get("result", {}).get("action")
        ]
        return dict(
            left=first,
            right=second,
            reused=second["reused"],
            changedCaseIds=changed,
            passed=first["passed"] and second["passed"],
        )
    finally:
        if left_moderator is None:
            await a.close()
        if right_moderator is None:
            await b.close()


def compare(
    left: PolicyConfig, right: PolicyConfig, cases: list[TestCase], **options
) -> ComparisonReport:
    return asyncio.run(compare_async(left, right, cases, **options))

File: packages/typescript/src/evaluation.ts
import {
  Moderator,
  ConfigurationError,
  ModerationError,
  decideResult,
  digest,
  inferenceHash,
  mockResponse,
  resolvePolicy,
  validateCases,
  validateSchema,
  type Policy,
  type PolicyConfig,
  type TestCase,
  type QualityGates,
  type ModerationResult,
  type Action,
} from "./core.js";
export interface CaseResult {
  id: string;
  expectedAction: Action;
  latencyMs: number;
  result?: ModerationResult;
  error?: { code: string; message: string };
}
export interface FilterMetrics {
  tp: number;
  fp: number;
  tn: number;
  fn: number;
  precision: number | null;
  recall: number | null;
  falsePositiveIds: string[];
  falseNegativeIds: string[];
}
export interface EvaluationReport {
  schemaVersion: 1;
  policyHash: string;
  policyVersion: string;
  inferenceHash: string;
  datasetHash: string;
  requestedModel: string;
  total: number;
  completed: number;
  errors: number;
  accuracy: number;
  reviewRate: number;
  confusion: Record<Action, Record<Action, number>>;
  filters: Record<string, FilterMetrics>;
  latencyMs: { mean: number; p95: number };
  usage: { input_tokens: number; output_tokens: number };
  failedCaseIds: string[];
  cases: CaseResult[];
  gateFailures: string[];
  passed: boolean;
  reused: boolean;
}
export function validateGates(gates: QualityGates, policy: Policy): void {
  validateSchema("gates", gates);
  for (const id of Object.keys(gates.filters ?? {}))
    if (!policy.filters.some((f) => f.id === id))
      throw new ConfigurationError(`Unknown gate filter: ${id}`);
}
export function summarize(
  policy: Policy,
  model: string,
  cases: TestCase[],
  rows: CaseResult[],
  gates: QualityGates = { minAccuracy: 1 },
  reused = false,
): EvaluationReport {
  validateGates(gates, policy);
  const actions: Action[] = ["allow", "review", "block"];
  const confusion = Object.fromEntries(
    actions.map((a) => [a, Object.fromEntries(actions.map((b) => [b, 0]))]),
  ) as EvaluationReport["confusion"];
  const filters = Object.fromEntries(
    policy.filters.map((f) => [
      f.id,
      {
        tp: 0,
        fp: 0,
        tn: 0,
        fn: 0,
        precision: null,
        recall: null,
        falsePositiveIds: [],
        falseNegativeIds: [],
      },
    ]),
  ) as Record<string, FilterMetrics>;
  const usage = { input_tokens: 0, output_tokens: 0 },
    failedCaseIds: string[] = [];
  let correct = 0,
    reviews = 0,
    errors = 0;
  rows.forEach((row, i) => {
    if (!row.result) {
      errors++;
      failedCaseIds.push(row.id);
      return;
    }
    const r = row.result;
    confusion[cases[i].expectedAction][r.action]++;
    if (r.action === cases[i].expectedAction) correct++;
    else failedCaseIds.push(row.id);
    if (r.action === "review") reviews++;
    usage.input_tokens += r.usage.input_tokens;
    usage.output_tokens += r.usage.output_tokens;
    for (const [id, truth] of Object.entries(cases[i].labels ?? {})) {
      const predicted = r.filters[id].action !== "allow",
        m = filters[id];
      if (predicted && truth) m.tp++;
      else if (predicted) {
        m.fp++;
        m.falsePositiveIds.push(row.id);
      } else if (truth) {
        m.fn++;
        m.falseNegativeIds.push(row.id);
      } else m.tn++;
      if (predicted !== truth && !failedCaseIds.includes(row.id))
        failedCaseIds.push(row.id);
    }
  });
  for (const m of Object.values(filters)) {
    m.precision = m.tp + m.fp ? m.tp / (m.tp + m.fp) : null;
    m.recall = m.tp + m.fn ? m.tp / (m.tp + m.fn) : null;
  }
  const durations = rows.map((r) => r.latencyMs).sort((a, b) => a - b);
  const accuracy = correct / cases.length,
    reviewRate = reviews / cases.length,
    gateFailures: string[] = [];
  if (gates.minAccuracy !== undefined && accuracy < gates.minAccuracy)
    gateFailures.push("minAccuracy");
  if (gates.maxReviewRate !== undefined && reviewRate > gates.maxReviewRate)
    gateFailures.push("maxReviewRate");
  for (const [id, gate] of Object.entries(gates.filters ?? {})) {
    for (const [key, metric] of [
      ["minPrecision", "precision"],
      ["minRecall", "recall"],
    ] as const) {
      if (
        gate[key] !== undefined &&
        (filters[id][metric] === null || filters[id][metric]! < gate[key]!)
      )
        gateFailures.push(`${id}.${key}`);
    }
  }
  return {
    schemaVersion: 1,
    policyHash: digest(policy),
    policyVersion: policy.version,
    inferenceHash: inferenceHash(policy, model),
    datasetHash: digest(cases),
    requestedModel: model,
    total: cases.length,
    completed: cases.length - errors,
    errors,
    accuracy,
    reviewRate,
    confusion,
    filters,
    latencyMs: {
      mean: durations.reduce((a, b) => a + b, 0) / rows.length,
      p95: durations[Math.ceil(durations.length * 0.95) - 1],
    },
    usage,
    failedCaseIds,
    cases: rows,
    gateFailures,
    passed: errors === 0 && gateFailures.length === 0,
    reused,
  };
}
export interface EvaluationOptions {
  moderator?: Moderator;
  concurrency?: number;
  gates?: QualityGates;
  signal?: AbortSignal;
}
export async function evaluate(
  policy: PolicyConfig,
  cases: TestCase[],
  options: EvaluationOptions = {},
): Promise<EvaluationReport> {
  const moderator = options.moderator ?? new Moderator({ policy });
  const resolved = resolvePolicy(policy);
  if (digest(resolved) !== digest(moderator.policy))
    throw new ConfigurationError(
      "Moderator and evaluation policies must match",
    );
  validateCases(cases, resolved);
  validateGates(options.gates ?? { minAccuracy: 1 }, resolved);
  const concurrency = options.concurrency ?? 4;
  if (!Number.isInteger(concurrency) || concurrency < 1 || concurrency > 64)
    throw new ConfigurationError("concurrency must be an integer from 1 to 64");
  const rows: CaseResult[] = new Array(cases.length);
  let next = 0;
  await Promise.all(
    Array.from({ length: Math.min(concurrency, cases.length) }, async () => {
      while (next < cases.length) {
        const index = next++,
          c = cases[index],
          start = performance.now();
        const row: CaseResult = {
          id: c.id,
          expectedAction: c.expectedAction,
          latencyMs: 0,
        };
        try {
          row.result = await moderator.moderate(c.message, c.history, {
            signal: options.signal,
          });
        } catch (error) {
          row.error = {
            code: error instanceof ModerationError ? error.code : "execution",
            message:
              error instanceof ModerationError
                ? error.message
                : "Evaluation failed",
          };
        }
        row.latencyMs = performance.now() - start;
        rows[index] = row;
      }
    }),
  );
  return summarize(resolved, moderator.model, cases, rows, options.gates);
}
/** Reuse only identical questions/model and dataset; a threshold change needs no inference. */
export function rethreshold(
  report: EvaluationReport,
  policy: PolicyConfig,
  cases: TestCase[],
  gates: QualityGates = { minAccuracy: 1 },
): EvaluationReport {
  const resolved = resolvePolicy(policy);
  validateCases(cases, resolved);
  if (
    report.inferenceHash !== inferenceHash(resolved, report.requestedModel) ||
    report.datasetHash !== digest(cases) ||
    report.cases.length !== cases.length
  )
    throw new ConfigurationError(
      "Replay requires identical questions, requested model, and dataset",
    );
  const rows = report.cases.map((row, i) => {
    if (row.id !== cases[i].id)
      throw new ConfigurationError("Replay case order mismatch");
    if (!row.result) return { ...row, latencyMs: 0 };
    const raw = mockResponse(
      Object.fromEntries(
        Object.entries(row.result.filters).map(([k, v]) => [k, v.probability]),
      ),
      row.result.model,
    );
    return {
      id: row.id,
      expectedAction: cases[i].expectedAction,
      latencyMs: 0,
      result: decideResult(resolved, raw),
    };
  });
  return summarize(resolved, report.requestedModel, cases, rows, gates, true);
}
export async function compare(
  left: PolicyConfig,
  right: PolicyConfig,
  cases: TestCase[],
  options: Omit<EvaluationOptions, "moderator"> & {
    leftModerator?: Moderator;
    rightModerator?: Moderator;
  } = {},
) {
  const a = options.leftModerator ?? new Moderator({ policy: left }),
    b = options.rightModerator ?? new Moderator({ policy: right });
  // Validate both before any billable request.
  const pa = resolvePolicy(left),
    pb = resolvePolicy(right);
  validateCases(cases, pa);
  validateCases(cases, pb);
  validateGates(options.gates ?? { minAccuracy: 1 }, pa);
  validateGates(options.gates ?? { minAccuracy: 1 }, pb);
  if (digest(a.policy) !== digest(pa) || digest(b.policy) !== digest(pb))
    throw new ConfigurationError(
      "Moderator and comparison policies must match",
    );
  const first = await evaluate(left, cases, { ...options, moderator: a });
  const second =
    inferenceHash(a.policy, a.model) === inferenceHash(b.policy, b.model)
      ? rethreshold(first, right, cases, options.gates)
      : await evaluate(right, cases, { ...options, moderator: b });
  return {
    left: first,
    right: second,
    reused: second.reused,
    changedCaseIds: cases
      .filter(
        (_, i) =>
          first.cases[i].result?.action !== second.cases[i].result?.action,
      )
      .map((c) => c.id),
    passed: first.passed && second.passed,
  };
}
